JavaScript2/Week5/review.md

```diff
@@ -11,4 +11,200 @@ this review covers:
+buttonElement.addEventListener('click', buttonClicked);
+```
+
+
+
+**Callbacks** 
+
+Now lets learn about callbacks!
+Well actually you have already made callbacks!
+When you give a function to an event listener or a timer or when fetching data you are using a callback function
+
+Lets create a callback function when someone writes in a input element
+```javascript
+const callback = function() {
+    console.log("Someone is writing!!");
+}
+
+document.querySelector('input').addEventListener('input', callback);
+```
+
+**Ajax Requests** 
+
+When us frontend developers want to get information about a logged in user,
+get search results from a search, check if someones passwors is correct
+we have to ask a server form these things. This communication between a client and 
+a server happens with using ajax requests.  
+
+```javascript
+const url = 'https://api.github.com/users/benna100';
+// Create new ajax call with the js function called XMLHttpRequest
+const req = new XMLHttpRequest();
+req.addEventListener('load', function () {
+    // This in here is our callback function
+    // Check our server responsecode, 200 means ok, success: https://en.wikipedia.org/wiki/List_of_HTTP_status_codes 
+    if (this.status === 200) {
+        const user = JSON.parse(req.responseText);
+        console.log(user);
+    } else {
+        console.log('Something is probably wrong with the url');
+    }
+});
+
+req.addEventListener('error', function () {
+    console.log('Server error like timeout');
+});
+```
+Initializes a request with an http method
+```
+req.open("GET", url);
+```
+Sends the request 
+```
+req.send();
+```
+
+Show the network tab in chrome and filtering for xhr
+
+
+**API - Application Programming Interface**
+
+What is an interface?
+Programmers can use an api to access complex functionality in a simple way. 
+Electricity socket. You just plug your appliance into the wall and it works.
+We dont have to worry about the wiring or anything. The complex functionality has been abstracted away.
+
+Examples of API's: 
+- DOM document.queryselector, 
+- Geolocation:navigator.geolocation.getCurrentPosition
+- accelerometer, 
+- facebook 
+- twitter
+
+**Web API**'s are like that, but just for getting data from a server. 
+There are some funny apis to play with:
+http://deckofcardsapi.com/
+http://api.giphy.com/v1/gifs/search?api_key=2FuF3E9nyFeXWt3aHIfkxtKTUGn73v0w&q=smile
+https://github.com/toddmotto/public-apis
+This incredibly unique service generates true random numbers by measuring quantum fluctuations of a vacuum in real-time!
+https://qrng.anu.edu.au/API/jsonI.php?length=1&type=uint8&#8217
+https://yesno.wtf/api/
+
+```javascript
+const api_key = '2FuF3E9nyFeXWt3aHIfkxtKTUGn73v0w';
+const query = 'smile'
+
+const giphyUrl = 'http://api.giphy.com/v1/gifs/search?api_key=' + api_key + '&q=' + query;
+
+// Create new ajax call with the js function called XMLHttpRequest
+
+const giphyReq = new XMLHttpRequest();
```

Comment:
The examples should really use fetch as XMLHttpRequest is outdated and it does not make sense teaching the students something that is not the recommended method.

https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API

Reply:
@kevinsimper Sounds like a reasonable idea - can you add it, or should it be the author? Benjamin is sailing for several days..

Reply:
Reason XMLHttpRequest is used is because they haven't introduced promises yet.  Which is fine, in this case.

Reply:
I agree that we should use fetch. But right now we use XMLHttpRequest as part of the callback learnings. If we were to not use XMLHttpRequest we need to introduce promises a lot earlier or we could when we introduce promises make them use the fetch api instead of XMLHttpRequest. On sunday i will talk about promises and then show them the fetch api.